cibuildwheel/linux.py

```diff
@@ -217,6 +217,18 @@ def build(options: BuildOptions) -> None:
                         # clean up test environment
                         docker.call(['rm', '-rf', venv_dir])
 
+                    existing_output_files = docker.call(['ls'], cwd=container_output_dir).split('\n')
+                    for repaired_wheel in repaired_wheels:
+                        if repaired_wheel.name in existing_output_files:
+                            message = f'Created wheel ({repaired_wheel}) already exists in output directory.'
+                            if 'abi3' in repaired_wheel.name:
+                                message += ('\n'
+                                            "It looks like you are building wheels against Python's limited API/stable ABI;\n"
+                                            'please limit your Python selection to a single version when building limited API\n'
+                                            'wheels, with CIBW_BUILD.')
+                            log.error(message)
+                            sys.exit(1)
```

Comment:
Do we want `sys.exit`, or an exception? Or downgrade this to just a warning (though Windows will error, anyway?).

And if `sys.exit`: just `1`, or a new/specific error code?

Reply:
How hard would it be to just skip the build here and go directly to tests when a wheel is already built? It should likely also be an error if tests are empty, as you should not "build" with multiple versions but not test? That would provide the "full support" solution.

Reply:
We've already done the build at this point. To do the 'full support' thing, we'd need to predict the produced wheel name without actually doing the build.

Reply:
The first time, you just build like normal. Then you could register that you produced a abi3 wheel, and future builds would skip building. For python-requires, you have to read the config, since you don't know what you are allowed to produce. For ABI3, you could just use the technique above and react to the built, rather than trying to compute what will happen first.

Reply:
True, that should be possible! The main question is how close this is to `cibuildwheel`'s other behavior. Most of the time, you have to explicitly opt-in to things and `cibuildwheel` won't say "ah, you probably meant ...".

On a more practical note: how would you decide which Python version to build on? I don't fully understand `Py_LIMITED_API`, but it seems to also include a "minimum version"?

Reply:
> how would you decide which Python version to build on?

Via Requires-Python and/or build selectors. `cibuildwheel` builds the oldest Python selected, and works it's way forward. Say it builds 2.7, normal wheel. 3.5, normal wheel. Then 3.6 is a stable ABI build. From then on, it just tests the existing wheel, 3.6, 3.7, 3.8, and 3.9 are just tests of the ABI wheel.

Reply:
This would even work if someone had special logic in setup.py for the limited_abi only working after some version of minimum required Python, while any other non-reactive method would not. You don't need to worry about the other direction (yet?), since the limited ABI works on all future Pythons (at least until Python 4).

Reply:
Right, OK, that could work, indeed :-)

Reply:
I like this idea, yeah! Avoiding the metadata read would be great.